ibis/expr/analytics.py

```diff
@@ -13,106 +13,86 @@
 # limitations under the License.
 
 
+import ibis.expr.rules as rlz
 import ibis.expr.datatypes as dt
-import ibis.expr.types as ir
-import ibis.expr.rules as rules
 import ibis.expr.operations as ops
 
 
-def _validate_closed(closed):
-    closed = closed.lower()
-    if closed not in {'left', 'right'}:
-        raise ValueError("closed must be 'left' or 'right'")
-    return closed
-
-
-class BucketLike(ir.ValueOp):
+class BucketLike(ops.ValueOp):
 
     @property
     def nbuckets(self):
         return None
 
     def output_type(self):
-        ctype = dt.Category(self.nbuckets)
-        return ctype.array_type()
+        dtype = dt.Category(self.nbuckets)
+        return dtype.array_type()
 
 
 class Bucket(BucketLike):
 
-    def __init__(self, arg, buckets, closed='left', close_extreme=True,
-                 include_under=False, include_over=False):
-        self.arg = arg
-        self.buckets = buckets
-        self.closed = _validate_closed(closed)
+    __slots__ = ('arg', 'buckets', 'closed', 'close_extreme', 'include_under',
+                 'include_over')
```

Comment:
@cpcloud I've added `slots` for each operations to support argument order on py2 too. While this is a bit more boilerplate, I wouldn't consider bad. 
I can use plain `__init__` methods as well as the counter trick we've discussed, in case of the latter we need to enforce calling the rules.

Which approach do You prefer? 
a. `__slot__` or `argnames` or ...?
b. counter trick and `rlz.any()` or `rlz.value(dt.any)`
c. plain but idiomatic `__init__` functions*?

* which is not not that bad, but `op.args` would require another order definition

Reply:
Based on gitter conversation we ended up with the following API:
```python
class StringFind(ValueOp):
    arg = Arg(rlz.string)
    start = Arg(rlz.integer, default=None)
    end = Arg(rlz.integer, default=None)
    return_ = Return(like=arg, dtype=dt.int64)
```

Reply:
This will alse make possible to support operation definition with python `typing`.